amaranth/lib/crc/__init__.py

```diff
@@ -0,0 +1,417 @@
+            w = crc.compute([0])
+            f.append([int(x) for x in reversed(f"{w:0{self.crc_width}b}")])
+        for i in range(self.data_width):
+            crc.initial_crc = 0
+            w = crc.compute([2 ** i])
+            g.append([int(x) for x in reversed(f"{w:0{self.crc_width}b}")])
+        return f, g
+
+    def __repr__(self):
+        return f"Parameters({self.algorithm()!r}, data_width={self.data_width})"
+
+
+class Processor(Elaboratable):
+    """
+    Cyclic redundancy check (CRC) processor module.
+
+    This module generates CRCs from an input data stream, which can be used
+    to validate an existing CRC or generate a new CRC. It is configured by
+    the :py:class:`Parameters` class, which can handle most forms of CRCs.
+    Refer to that class's documentation for a description of the parameters.
+
+    The CRC value is updated on any clock cycle where ``valid`` is asserted,
+    with the updated value available on the ``crc`` output on the subsequent
+    clock cycle. The latency is therefore one clock cycle, and the throughput
+    is one data word per clock cycle.
+
+    The CRC is reset to its initial value whenever ``start`` is asserted.
+    ``start`` and ``valid`` may be asserted on the same clock cycle, in which
+    case a new CRC computation is started with the current value of ``data``.
+
+    With ``data_width=1``, a classic bit-serial CRC is implemented for the
+    given polynomial in a Galois-type shift register. For larger values of
+    ``data_width``, a similar architecture computes every new bit of the
+    CRC in parallel.
+
+    The ``match_detected`` output may be used to validate data with a trailing
+    CRC (also known as a codeword). If the most recently processed word(s) form
+    the valid CRC of all the previous data since ``start`` was asserted, the
+    CRC register will always take on a fixed value known as the residue.  The
+    ``match_detected`` output indicates whether the CRC register currently
+    contains this residue.
+
+    Parameters
+    ----------
+    parameters : Parameters
+        CRC parameters.
+
+    Attributes
+    ----------
+    start : Signal(), in
+        Assert to indicate the start of a CRC computation, re-initialising
+        the CRC register to the initial value. May be asserted simultaneously
+        with ``valid`` or by itself.
+    data : Signal(data_width), in
+        Data word to add to CRC when ``valid`` is asserted.
+    valid : Signal(), in
+        Assert when ``data`` is valid to add the data word to the CRC.
+    crc : Signal(crc_width), out
+        Registered CRC output value, updated one clock cycle after ``valid``
+        becomes asserted.
+    match_detected : Signal(), out
+        Asserted if the current CRC value indicates a valid codeword has been
+        received.
+    """
+    def __init__(self, parameters):
+        if not isinstance(parameters, Parameters):
+            raise TypeError("Parameters must be of type Parameters")
```

Comment:
Could you do this the same way we handle runtime type checks elsewhere, please?
```suggestion
            raise TypeError("Algorithm parameters must be of type Parameters, "
                            "not {!r}"
                            .format(parameters))
```